Add Automated updated of `autocomplete-css` `completions.json`

confused-Techie opened the pull request:
Alright so this PR turned into a lot more than I had intended it to be.

In short this PR allows `autocomplete-css` to continue to update it's `completions.json` as well as getting it updated to today's standards. This was done by completely rewriting the `update.coffee` and now using `update.js`

That's the short version, but for some context:

---

Previously `autocomplete-css` would use this JSON file located `https://raw.githubusercontent.com/adobe/brackets/master/src/extensions/default/CSSCodeHints/CSSProperties.json` to update it's completions. This file would provide it with CSS Properties, which it would then manually query MDN for each property, and scrape the webpage for relevant data to build a description.

But as one might imagine this file became out of date, and no longer would provide us with the newest CSS Properties. So rather than try to find someone else that was manually managing this data, I turned to a resource that would keep this data updated for some time, and had at least some reason to keep things backwards compatible when inevitably the data structure changes, hopefully it won't be for some time.

But I ended up using `@webref/css` [NPM](https://www.npmjs.com/package/@webref/css?activeTab=readme) [GitHub](https://github.com/w3c/webref) which as they define the standards, they will continue to have the most up to date information we could get. This package provides an easy way to get an abundance of data about every possible specification of CSS. More than we needed, but this would give us what we need.

The two big hurdles:

1) The data format here was slightly tricky, requiring a bit of code to work around it, but nothing to crazy. The real difficult part was their usage of [CSS Value Definition Syntax](https://developer.mozilla.org/en-US/docs/Web/CSS/Value_definition_syntax), which while obvious that they would use it, meant we couldn't easily grab the data needed for our `completions.json`, that is grab the valid values that each property would accept. Instead of trying to scrape MDN or anything I instead built a super simple parser. This parser is located in `cssValueDefinitionSyntaxExtractor.js` (A bit obvious and long I know) but this will go ahead and accept any CSS Value Definition Syntax String and parse out **all** valid values of a property. While this does ignore many aspects of the syntax, such as some values are only valid when another is not present, that was out of the scope of what we needed. And would require significant change on the `autocomplete-css` package as a whole, even if in the future I'd love to see this being able to utilize the syntax directly and actually provide autocompletions based on that.

2) The second challenge here was to actually then get descriptions of our properties. This was done utilizing [`mdn/content`](https://github.com/mdn/content) which is the GitHub repo that contains the Markdown documents of all of MDN's website. With this we are much more easily able to parse out the needed descriptions from the Markdown rather than having to worry about hitting their website dozens of times and dealing with the overhead of that. A quick note, is these documents are also the same way we are collecting our HTML tags for our `completions.json` file so those will continue to be updated.

The last note here, our `pseudoSelectors` are not updated. At this time I was unable to identify the best way to keep these updated, so for the time being we are still using the same ones that we had previously. There is a function that can be used to start updating them, once we determine the best way. But for now that is not being accomplished.

But with all that said, our `pseudoSelectors` have not been changed, but our `properties` are now as up to date as the current CSS spec is, and `tags` are as updated as the MDN website is. So I'd like to think this is a huge improvement over what we previously had, even if I now have ideas for how to improve our package as a whole.

---

Resolves #393

savetheclocktower commented on packages/autocomplete-css/cssValueDefinitionSyntaxExtractor.js:
```diff
@@ -0,0 +1,163 @@
+      "!": "Exclamation Point Multiplier"
+    };
+
+    this.seperators = {
```
```suggestion
    this.separators = {
```

confused-Techie replied:
Ahh whoops, lemme fix that where needed

DeeDeeG commented:
This is very impressive, I didn't get a chance to review all the JS of the updater script or anything like that, but the spirit of this seems spot on.

Would love to take a closer look if I have time, but it seems pretty in-depth.

Lite approve ✅, I don't mind this merging, just wish I could be more thorough in reviewing it at the moment!

confused-Techie commented:
> This is very impressive, I didn't get a chance to review all the JS of the updater script or anything like that, but the spirit of this seems spot on.
> 
> Would love to take a closer look if I have time, but it seems pretty in-depth.
> 
> Lite approve ✅, I don't mind this merging, just wish I could be more thorough in reviewing it at the moment!

Appreciate it! Feel free to ask any questions about it that you are curious about. This one did end up being way more complicated than originally intended

icecream17 commented on packages/autocomplete-css/completions.json:
```diff
@@ -1,2087 +1,4381 @@
+    "footer",
+    "form",
+    "frame",
+    "frameset",
+    "head",
```
the tags `h1` to `h6` got removed. (no other tags got removed)

icecream17 commented on packages/autocomplete-css/completions.json:
```diff
@@ -1,2087 +1,4381 @@
+    "footer",
+    "form",
+    "frame",
+    "frameset",
+    "head",
```
```suggestion
    "frameset",
    "h1",
    "h2",
    "h3",
    "h4",
    "h5",
    "h6",
    "head",
```

these were the only tags removed

confused-Techie replied:
Thanks for noticing this. Although we can't just add them back into the `completions.json` as that data would then be overridden the next time the update script is used.

It looks like the page from MDN on these tags, located [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements) are actually within a folder called `heading_elements` on their docs repo. Which we use the title of these folders to create the tag list.

So some logic has to be added in to validate all elements added via the tags method like this, and replace things like `heading_elements` with the proper `h1->h6` tags.

It might be worth it to look through the [other folders listed here](https://github.com/mdn/content/tree/main/files/en-us/web/html/element), and see if there's any cases of the same thing happening that we haven't noticed.

confused-Techie replied:
Alright @icecream17 I've added a small check that allows us to handle this. As well as making it easy to expand in the future if we find this has happened again. Searching through the MDN docs website, I couldn't find any other occurrences of this currently happening.

And I've also of course ran the script again to test the changes and confirm we now get `h1->h6` as well as remove `heading_elements`

DeeDeeG commented on packages/autocomplete-css/completions.json:
```diff
@@ -1,2087 +1,4381 @@
+      ],
+      "description": ""
+    },
+    "display": {
+      "values": [],
+      "description": "The display CSS property sets whether an element is treated as a block or inline element and the layout used for its children, such as flow layout, grid or flex."
+    },
```
This used to have values populated for it, such as [`block`, `flex`, `grid`, `none`] ...

Wonder why the values didn't get picked in the automated update here. Hmmmmm.

confused-Techie replied:
It seems that the actual spec has the following CSS Value Syntax of `<display-outside> || [ <display-inside> | math ]`.

So I'd reason to assume that it maybe can't find those other Value Data Types, although that doesn't explain why `math` doesn't show up. As that's just a value literal. Hm...

confused-Techie replied:
Wait I take that back.

There's two `display` definitions.
- The first located within [`css-display.json`](https://github.com/w3c/webref/blob/main/ed/css/css-display.json)
- The second located within [`mathml-core.json`](https://github.com/w3c/webref/blob/main/ed/css/mathml-core.json)

These two items define the following CSS Value Definition Syntax: (Respectively)

- `[ <display-outside> || <display-inside> ] | <display-listitem> | <display-internal> | <display-box> | <display-legacy>`
- `<display-outside> || [ <display-inside> | math ]`

The thing is, is the `mathml-core` spec defines this under `newValues` whereas `css-display` spec defines under just `value`.

So what happens is we first encounter the `css-display` spec, and parse it's values. Which seems we can't find the referred to value groups, which means we then only end up with an empty array.  Then we hit the `mathml-core` spec which because the `value` key doesn't exist, only `newValue` we get `undefined` and again receive an empty array. 

This is a little disappointing because according to `w3c/webref` (which is who provides `webref/css`) [`No duplicate top-level definitions or members.`](https://github.com/w3c/webref) is one of their "Curation guarantees". So I'm surprised we are finding a duplicate definition here. Unless I've misread their statement.

Otherwise the question then comes down to why we aren't finding the value group definitions? It's possible they may not exist within the document how we parse it, And I'm still looking into where that definition may be, if it exists.

---

But a quick confirmation on this concern can be seen here, adding the following to `buildProperties()` within it's second level loop I added 

```javascript
if (prop.name === "display") {
  console.log(prop.name);
  console.log(prop.value);
  console.log(propValues);
}
```

Which will output the following:

```bash
PS D:\pulsar-edit\pulsar\packages\autocomplete-css> node update.js
display
[ <display-outside> || <display-inside> ] | <display-listitem> | <display-internal> | <display-box> | <display-legacy>
[]
display
undefined
[]
Updated all `autocomplete-css` completions.
```

confused-Techie replied:
@DeeDeeG just to make sure you're able to see it, if you've got any ideas or input

confused-Techie replied:
So just figured it out. We were passing a single spec `values` object over to `getValuesOfProp()` which then expected to receive the full `css` object with all specs. So it actually seems that no single Value Group was able to be properly resolved. So I've just fixed that and pushing it now

confused-Techie replied:
Yeah look at this [diff](https://github.com/pulsar-edit/pulsar/pull/398/commits/705e0d9a0d9172aa59a1fc42728ce97078ae5f92). There's all our missing values

DeeDeeG replied:
Wow, that got a _lot_ of values on a lot of different CSS properties. Love to see it!

confused-Techie commented:
So with all the changes in, and all of the first issues people noticed having been resolved. I've moved on to looking at the tests: Which are not what we are hoping for:

* autocomplete-css: 125 Failures - We expect 0
* find-and-replace: 66 Failures - Tad High but within expected range
* symbols-view: 2 Failures - Expected
* tree-view: 2 Failures - Expected

So obviously we have many failures within the tests of `autocomplete-plus` that we do not expect. To be frank I haven't run tests, as the structure of the `completions.json` hasn't changed, so there should be zero change in how the tests run. That is unless of course there's any manual checking for values suggested or returned that no longer match up with the updated data. So I'll go ahead and continue to take a look at this.

DeeDeeG commented:
Some of it, IIRC from trying similar a while ago, was a hard-coded expected number of properties or values or some such assertion. That'll need updating.

Perhaps various other checks assume old values/properties.

confused-Techie commented:
> Some of it, IIRC from trying similar a while ago, was a hard-coded expected number of properties or values or some such assertion. That'll need updating.
> 
> Perhaps various other checks assume old values/properties.

Yeah, it's a little strange. Essentially the vast majority that are failing are expecting the top level completions that would be returned. For example `border-` would expect the first result to be `border:` and the second result to be `border-radius`

Which honestly checks out. This makes sense that you would expect this rather than, what we are currently getting, `background-blend-mode` and `background-position`. Thing is, obviously I didn't change any handling of this, which to me means one of two things. The `completions.json` is expected to be ordered in some such way, since the values returned are truly just the top items that begin with a `b`, or that the end all be all of our completions relies on the function `firstCharsEqual()` within `provider.coffee`. Which that does exactly what it sounds like. Compares if the first characters match between a completion candidate and the buffer text. Either way though that may mean I need to apply some fixes to `autocomplete-css`.

But honestly as I don't know coffeescript, or the existing code here, I'm hoping the change could be simple. Possible as simple as ordering the completions before being passed back to the user manually, such as I suggested on the linked issue in my PR description using LCS to rank the results then return them. Although I'd honestly much rather see a change like that land in `autocomplete-plus` instead of here. But I'll keep trying to determine what exactly has gone wrong

DeeDeeG commented:
Not sure exactly what the "ordered completions" file was about, that they went through the trouble of referring to usage statistics and place more popularly used properties/values first, based on crawling the actual websites of the web and inspecting the properties/values used most.

This PR deletes those files, so I was going to assume they had only been used for generating the list of completions, and yet the ordering bit feels like it would be unnecessary in that case... So I'm not sure where, but I get a feeling the "ordered completions" file(s) may still be needed... ?? Or some ordering they provided as a middle step before ordering the final output is important?

confused-Techie commented:
@DeeDeeG Just wow. Yeah looking back through the source, and `sorted-property-names.json` I think this is exactly what is happening. 

The order of `sorted-property-names.json` 

1) Matches the order for the completions order that's expected within the tests 
2) Is the canonical order that the `completions.json` would be written in.

A side note that is just unbelievable, seemingly if a property was retrieved via the API that didn't exist within `sorted-property-names.json` it would be discarded. Meaning that Atom had always had a static list of accepted properties. So the old update method makes very little sense at this point.

But what that does mean, is the semi-random order of `sorted-property-names.json` is important, and is causing the specs to fail. But like you said, how in the world did they order it? I refuse to believe it was done based on usage or popularity, since that is pure insanity. I'm hoping rather it's built according to the way `autocomplete-plus` calculates autocompletions. Such as to be the most streamlined method. But I'm not to sure.

---

An interesting note. I tried running the new completions.json file locally, and I couldn't replicate the order being retrieved via the specs manually at all.

So I thought, maybe `autocomplete-plus` does have some code that attempts to order the suggested completions, so activating that package as well in our specs does get 10 or so more tests passing now. From `125` to `116` iirc.

But yeah otherwise that's just a mess of behavior. 

I'm wondering how poorly performance would be if we stop just using `firstCharsEqual()` since to me it seems that's whats trigger the extra completions. But I honestly might want to implement some LCS functionality here, and try to test if performance is impacted. But that was something I'd at least hoped to talk in another PR. Since otherwise I feel I'd have to match the order, or change the specs to expect what's now being returned.

DeeDeeG commented:
> `sorted-property-names.json` is updated manually - take a look at https://developer.microsoft.com/en-us/microsoft-edge/platform/usage/ and https://www.chromestatus.com/metrics/css/popularity for guidance.

From the [autocomplete-css README.md](https://github.com/pulsar-edit/pulsar/tree/b697a51fa7c5dd7988d8d081b1e87f041d8f5eb1/packages/autocomplete-css#readme).

confused-Techie commented:
> > `sorted-property-names.json` is updated manually - take a look at https://developer.microsoft.com/en-us/microsoft-edge/platform/usage/ and https://www.chromestatus.com/metrics/css/popularity for guidance.
> 
> From the [autocomplete-css README.md](https://github.com/pulsar-edit/pulsar/tree/b697a51fa7c5dd7988d8d081b1e87f041d8f5eb1/packages/autocomplete-css#readme).

Wow, when you look everywhere but the readme. Thanks for the link, but that's just gross. Gotta find a solution to that one way or another

icecream17 commented on packages/autocomplete-css/update.js:
```diff
@@ -0,0 +1,300 @@
+  // within their docs. By simply grabbing all folders of tag docs by their name
+
+  // Some of the page titles from MDN's docs don't accurately reflect what we
+  // would expect to apper. The object below is named after what the name of the
```
```suggestion
  // would expect to appear. The object below is named after what the name of the
```
random typo heh. makes me think there's probably a typo-comment-linter out there, though of course there's probably going to be a lot of false positives...

confused-Techie commented:
Alright, all tests are now finally passing locally on my machine.

Essentially the tests for `autocomplete-plus` were mostly a validation of the `completions.json` file and it's order and it's properties. Where nearly every single test would test the amount of completions returned for any property, which of course now is larger. Where they would test the exact order of completions provided, which of course has no changed. Or some even tested the text of the description provided by a completion, by of course checking it's exact index within the completions array provided.

So with that said there's a few things I've done. Unfortunately this makes the tests less readable when they fail, but does mean they will fail significantly less often.

In the case of checking the length of completions that would be provided for a character, since of course now there are more completions every single check failed expecting a smaller number than we now have. While there was a whole discussion about the best choice here over on [Discord](https://discord.com/channels/992103415163396136/992103415163396139/1078965950398869534) in the end I've opted to do the simplest thing that can hopefully protect against regression. Checked if the amount of completions we have is *higher* than what was previously expected. This at least ensures the test will pass, that is until they start removing items from the CSS spec faster than they add them. Which I don't personally think is something we need to be super worried about.

The other options discussed where ensuring the amount is over 0, or seeing if it's within a certain range. But in the end I do personally feel this is the best solution.

But now for the case where the tests would check the exact order of completions.

By this I mean there were endless statements like `expect(completions[0].text).toBe 'display: '` Which the thing is, the reason this mostly worked was because of `sorted-property-names.json`. Apparently previously the Atom team had used a manually made list of completions, that was taken from Microsoft Edge web usage statistics and would rank the properties in order of popularity. And that is how they would then test, to ensure items were still ranked in terms of popularity. Thing is, not sure if this wasn't a feature at the time, but from my understanding and experience is `autocomplete-plus` actually helps to order the completions according to what most closely matches what a user typed, and beyond that I don't think we actually care about the list returned. If we do, or if it's a problem that when a user types just `d` and they don't get back the most commonly used items of the language we could revisit this, but I personally think that entire system is prone to failure, and way more trouble than it's worth.

So in that case I've rewritten the majority of tests (whatever was failing) to stop doing this. From what I rewrote we no longer expect any exact order. All we do is check if the value we want to test for exists somewhere in the completions. This is done with `isValueInCompletions()` at the top of the spec file. And using it we can just check if `display: ` exists in one of the completions provided, rather than checking is it the first completion provided. Granted if values are removed or the spec changes drastically this could mean aspects here have to be rewritten, but for the most part I think this is a safe compromise of ensuring we have the values we expect, but allowing the `completions.json` file to change over time.

I will admit I hadn't wanted to touch specs on this PR, and assumed I wouldn't have to. But from here the last thing to do is remove the note about `sorted-property-list.json` within the readme and wait for all tests to pass, which like I said it does on my machine.

---

Side Note: One last thing I had to add was inserting the implicitly defined CSS Values onto all properties, this did cause our `completions.json` file to become much larger, but got many values that we tests for back into the file. Such as `inherit`

DeeDeeG commented on packages/autocomplete-css/spec/provider-spec.coffee:
```diff
@@ -42,6 +42,12 @@ describe "CSS property name and value autocompletions", ->
       activatedManually: options.activatedManually ? true
     provider.getSuggestions(request)
 
+  isValueInCompletions = (value, array) ->
+    result = []
+    for i in array
+      result.push(i.text)
+    return value in result
+
```
This function looks helpful!

I think naming the input variable "`array`" more contextually specific as "`completions`" would help readability: 

```suggestion
  isValueInCompletions = (value, completions) ->
    result = []
    for i in completions
      result.push(i.text)
    return value in result
    
```

I might even rename "`result`" to "`completionsTextNodes`" or "`completionsText`" or something, but it's less important and I'm not sure if everyone agrees with my tendency to be really really explicit and literal and at times verbose with what my variable names are. I just don't trust myself to remember what I was doing when I wrote code a week ago/yesterday, TBH.

But I think it's somewhat clear after reading this short function that `result` is just a temporary variable to hold an array with some sub-property `text` that we care about extracted from a more complicated `completions` data structure/array.

confused-Techie replied:
You do have a really valid point about making code obvious at a glance. So I think I'll go ahead and make the changes you are suggesting here. It is small, but important for making things maintainable. But also small enough I have a hard time seeing myself looping back to this, at least until the next time someone is updating these specs (or packages).

So I really do appreciate the in depth review, and I'll get on this one

confused-Techie replied:
Alright @DeeDeeG done in 7a746a76969385a672e159acc4a0572bf9bc3e93

DeeDeeG commented:
Specs are passing! I'm gonna take a closer look and test some things locally, but that's a pretty good vindication of this PR!

(Yeah, it's unfortunate to have to update specs to get any changes to the completions passing in CI -- BUT, BIG BUT, and I think this overrides that -- _not as unfortunate as the specs having been this brittle-ly hard-coded in the first place._)

I assume there was some effort before to make sure that, all else being equaol, popular CSS properties would have been the first to be suggested. Such as if you've typed a few characters and several equally-similar fuzzy suggestions exist for that input string.

We might endeavor to re-create that popularity-fist aspect, but IMO having all the completions _be there_, in whatever, is more important.

(This is almost a review, I just wanna test things. So if I do a formal UI-based "Review", I will probably just point back to this for my reasoning.)

DeeDeeG commented on packages/autocomplete-css/completions.json:
```diff
@@ -1,2087 +1,7441 @@
   ],
   "properties": {
-    "width": {
+    "-webkit-text-fill-color": {
```
Just noting that it doesn't seem possible to autocomplete typing the _name_ of a prefixed property (one starting with `-webkit` or something like that), but their _values_ can still be autocompleted once you've typed the property _name_ out manually.

I dunno what it would take to fix that, but at least it'd be a "known bug" now if we do ship it like this.

confused-Techie replied:
Yeah same, once we get into the language portions of the autocomplete functionality, I wouldn't totally know where to start. So this is good to know, if at least, like you said for a known bug.

DeeDeeG commented on packages/autocomplete-css/completions.json:
```diff
@@ -1,2087 +1,7441 @@
+      ],
+      "description": "The animation-iteration-count CSS property sets the number of times an animation sequence should be played before stopping."
+    },
+    "animation-direction": {
+      "values": [
+        "inherit",
+        "initial",
+        "unset",
+        "normal",
+        "reverse",
+        "alternate",
+        "alternate-reverse"
+      ],
+      "description": "The animation-direction CSS property sets whether an animation should play forward, backward, or alternate back and forth between playing the sequence forward and backward."
+    },
```
I wonder if it would be possible to put the "inherit", "initial", "unset" at the _bottom_ so you don't have to look past or scroll past them to get to the "juicier" values (in my subjective opinion it'd be awesome to be able to autofill these but not see them prominently/see them last.)

Calling this one peroperty out since I think it's a good illustration, "normal" and "reverse" seem the more frequently/generally useful options to pick over "inherit" or "initial" and so on, but this is a general comment that applies to anything with "inherit", "initial", etc. possible values.

savetheclocktower replied:
Agreed — I think those three should always be at the bottom of the list of choices. I know it's getting into the weeds if we try to sort each individual property's possible options by frequency of usage, but those three are always present as possible values and hardly ever used.

confused-Techie replied:
So with this, the sorting may be difficult but I can look into it. 
The other option, do we want to exclude those three from autocompletions? Since that might technically be easier to do.

From memory, what's happening here, essentially the `interface` definitions we are getting from Chromiums DevTools all match to an element we get from W3C. 

So essentially we match the `interface` and that gives us so many properties, but each `interface` inherits another and can have several `includes` specified on them. The way we resolve this is by resolving each inclusion or inheritance, and each additional one we include or inherit can also include or inherit from there, but I'm pretty sure these three properties, are coming from something like the `interface` `HTMLhtml` which every single item eventually has an inheritance to. So we could just stop including that inheritance, or otherwise I can try to figure out how we can put that one at the bottom of the list. 

That might look more just like a manual filter on the finalized arrays than anything

DeeDeeG replied:
I got the implicit values "inherit", "initial", "unset" at the bottom, posted to a branch at my fork:
https://github.com/pulsar-edit/pulsar/compare/update-autocomplete-css...DeeDeeG:pulsar:update-autocomplete-css-implicit-last

<details><summary>Comparison permalink</summary>

Permalink (SHA-based) version of the same comparison: https://github.com/DeeDeeG/pulsar/compare/7a746a76969385a672e159acc4a0572bf9bc3e93...ebdd711d6481064a500fbdb27cbe6c4eb24675b2

</details>

@confused-Techie Would you be interested in me pushing these changes (commits) to this PR branch? Or might you be interested in it as a follow-up PR?

(I do personally think it'd be nicer to have the implicit values at the bottom, so I'm hoping to get this into the package some way or other.)

<details><summary>Note about unrelated/not strictly minimal changes in the diff of my branch (click to expand):</summary>

Running `node update.js` changed the order of some properties, but I don't _think_ it changed the values.

Also: (Again, not sure why, but) running `npm install` changed the formatting of the github URL on one line of `package-lock.json`, whether I used latest npm 8 or 9 (couldn't use npm 6 without major lockfile churn since it's lockfile v2), so I committed that. I always hate when lockfiles change after every install, they should be stable, so I committed that.

</details>

confused-Techie replied:
@DeeDeeG the changes you made are fantastic! And seems I had forgotten that those values were defined directly rather than being properly resolved, so this change is possible but also super simple. And yeah I'm not to worried about lock file changes tbh, since they do seem mostly random at one point and I'll just take it as that's what NPM needs to function.

As for the order of properties resolving that doesn't worry me to much. It seems very dependent on how files were read from disk, and what was read faster than other's when running the `css.listAll()` function, so I'd expect them to change during each run, for their ordering at least, of course not actual values.

But for your changes lets get those merged, and then I'll do a small edit for the bad check I have that you mentioned below!

DeeDeeG reviewed (commented):
Just a couple of things I wanted to point out for discussion?

Also, I have no specific line of code to comment this on, so I'll say it here: I've noticed the cursor can get stuck and stop my attempts to type or backspace, on this PR's binary sometimes, might be specific to Command+Tabbing (window switching) back and forth between Pulsar and the browser, with this PR's binary? Whereas I don't think I can reproduce this sticking/freezing on 1.102.0 Regular.

**EDIT**: Actually, I can reproduce this on 1.102.0 too. It happens after clicking out to the docs for a property --> link opens in browser, alt-tab/cmd-tab back to Pulsar, try to type or backspace around and cursor gets stuck soon. Reproducing on macOS 10.15, with this binary or 1.102.0 Regular.

~~Kind of makes me wonder if there's some empty array or object somewhere in `completions.json` that the code is tripping over? Or just the increased completions size making it hitch? I dunno, just guesses.~~

Overall, this is looking really good, though. Love that we have so many more properties/values that can be auto-completed. Any new properties from the past several years that would have been missing without this PR.

confused-Techie commented:
@DeeDeeG yeah Pulsar getting stuck sometimes doesn't sound at all ideal.

Doing a quick look around I didn't see any empty `values` arrays, but it's good to note that the original `completions.json` also had those.

If I had to be worried, I'd say it could be the increase `completions.json` filesize. 

Since for example before my PR it was 2,225 lines, whereas my PR makes the `completions.json` 7,579 lines. So this is a really big increase for a file that's essentially in memory. So yeah not the most encouraging thing

DeeDeeG commented:
Okay, I figured out the steps to reproduce.

- Start typing a property
- Click the "More.." link to read docs on MDN
- alt-tab/cmd-tab back to Pulsar
- Start typing or backspacing again

Unexpected result: typing and/or backspacing will stop working until you re-focus (click into) the CSS file again.

Happens for me on 1.102.0 Regular _or_ this PR's binary from Cirrus, macOS 10.15.

confused-Techie commented:
@DeeDeeG thanks for testing this so well!
Glad to see it isn't a problem I've created. But is good to know that this problem exists

DeeDeeG commented on packages/autocomplete-css/update.js:
```diff
@@ -0,0 +1,301 @@
+  // We will first insert the implicitly defined keywords that apply to all CSS properties
+  let values = [ "inherit", "initial", "unset" ];
+
+  if (!value || value.length < 0) {
```
Quick question, although it doesn't matter a ton honestly:

Is this `value.length < 0` a real condition that would happen?

<details><summary>Guesses (click to expand):</summary>

Is this `value` an object with a property on it called `length` that can be set to a number below zero, some negative value for `length`?

If not: Is it possible for some random JavaScript thing (an array or a string, I'm thinking?) to have a `.length` that is `< 0`? I don't think so...

</details>

Just a little puzzling.

I got it so the implicit values would all be last, locally, but ran into this bit and got a little confused.

DeeDeeG replied:
P.S. I promise I'm almost done reviewing this PR!! Thanks for all of it! It's pretty big but I like what I see so far! And am starting to lean toward "lets just ship it", though I will be proposing the "make the implicit values show last" change I'm working on soon. Ideally after the above question is resolved, but I can ignore this strange-looking check and leave it as-is honestly. It works, that's what matters.

confused-Techie replied:
Yeah this bit really is trying to protect from the fact that the `props` from Chromium's DevTools can be a zero length array or not exist (iirc).

So I think this was an incorrectly created attempt at checking 

```javascript
if (!value || value.length === 0) {}
```

So it likely was intended to be that. Although due to type coercion an array of `length` 0 will actually be a falsy value. So we could absolutely change this to just `if (!value) {}` with nothing else.

Good catch on some sloppy but functional code. But only functional since the `||` is there, since that second argument will always be true, except in the case where the array is `0` or non-existent, in which case the first check catches it.

Must've been sleepy or modified from a previous check that failed.

DeeDeeG reviewed (approved):
Thank you for bearing with review on this one!

I am approving because the improvements are really meaningful (lots of new properties/values! We can actually easily update this ongoing, as the new update script isn't a dead end anymore like the old one got to be!).

And any downsides are minor or negligible that we can easily live with (such as the ordering not being by popularity anymore, not a big deal honestly as long as we have the properties).

I can address the implicit values "inherit" "initial" "unset" being first via some commits I pushed to my fork, willing to do those as a follow-up PR at this point just to get this one done.

Again, thank you for your patience! Big contribution, thank you very much!

confused-Techie commented:
@DeeDeeG I really do appreciate you giving this all a review! I know this was a larger PR so taking your time is a good thing.

But seriously, you're the best!

What do you say we take care of you're changes for moving implicitly CSS properties to the end of props arrays, and take care of my bad check then get this guy merged with the above approval if nothing else comes about it?

DeeDeeG commented:
@confused-Techie I have pushed the aforementioned commits for putting "inherit", "initial", and "unset" at the _bottom_ of the values list for each property. They're in this PR now, I think from your comment above that that's what you were looking for. 👍

confused-Techie commented on packages/autocomplete-css/update.js:
```diff
@@ -0,0 +1,301 @@
+  // We will first insert the implicitly defined keywords that apply to all CSS properties
+  let values = [ "inherit", "initial", "unset" ];
+
+  if (!value || value.length < 0) {
```
```suggestion
  if (!value) {
```

Remove bad check

confused-Techie commented:
Awesome @DeeDeeG, I've also added in those changes for the final check.

And assuming all tests pass as we would expect, if you're approval still applies I can go ahead and merge this one if that sounds good to you?

DeeDeeG reviewed (approved):
Approving again, see my previous approval(s?) for more details why this PR is good more generally.

Re-approving because the simplified `if (!value)` check has landed, I ran `node update.js` with that change and my `completions.json` stayed stable, so it seems to have had no affect on the output, which is what I was hoping for.

Hoping to see CI within our expected failure rates and then I think this one is good to go!

DeeDeeG commented:
So, I think my changes to the completions must have done something not so great? Maybe gotten rid of some completions??

From the `autocomplete-css` specs, there's four of these:

```
CSS property name and value autocompletions
  CSS files
    it autocompletes property names with a prefix
      Expected 0 to be greater than 0.
        at jasmine.Spec.<anonymous> (/home/runner/work/pulsar/pulsar/node_modules/autocomplete-css/spec/provider-spec.coffee:121:51)
```

confused-Techie commented:
Thanks for the heads up @DeeDeeG I'll take a look.

Likely either my bad check was actually doing something, or that some properties are no longer getting the implicitly defined values.

DeeDeeG commented:
**UPDATE**: Nah, my changes just moved the `d`Property (Is this a real thing?) first when typing `d`, and the `d` property (not a real thing???????????) has no description, so the description length is in fact not greater than zero.

Another rather brittle test, thinking what we wanna do about it since the original is so hyper-specific.

Maybe have it type more characters, maybe a property we know is real and should have a description, such as `display`?

confused-Techie commented:
@DeeDeeG Ahh thanks for finding the root cause of the issue.

But yeah the `d` property is a valid query selector as it's part of the [SVG Spec](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d#using_d_as_a_css_property).

The reason our description isn't getting picked up is because it exists under MDN's SVG docs, rather than CSS docs.

Wonder if some of the other missing values exist there as well. 

May be worth adding that as a secondary docs location to collect descriptions, the same as has been done over on #405

DeeDeeG commented:
Note: We also have some draft properties like `ruby-merge` from somewhere.

https://drafts.csswg.org/css-ruby-1/#collapsed-ruby

This didn't have a `description` in the `completions.json` either, beyon the  empty string `""`.

confused-Techie commented:
> Note: We also have some draft properties like `ruby-merge` from somewhere.
> 
> https://drafts.csswg.org/css-ruby-1/#collapsed-ruby
> 
> This didn't have a `description` in the `completions.json` either, beyon the empty string `""`.

Yeah looking at all the empty descriptions, and have gotten nearly all of them solved locally, just taking a look at a last few ones, but should be up within 30 minutes hopefully

DeeDeeG commented on packages/autocomplete-css/completions.json:
```diff
@@ -3992,7 +3711,7 @@
         "unset"
       ],
-      "description": ""
+      "description": "« SVG Attribute reference home"
```
> ` "description": "« SVG Attribute reference home"`

Hmmmmmmmmm.

confused-Techie replied:
Wow faster than I could write about my update. Lemme look for that

confused-Techie replied:
Alright, resolved this one with the recent change

DeeDeeG commented on packages/autocomplete-css/completions.json:
```diff
@@ -6839,30 +6558,75 @@
         "unset"
       ],
-      "description": ""
+      "description": "> Warning: This property has been replaced by the break-after property."
```
> `"description": "> Warning: This property has been replaced by the break-after property."`

I guess if this description is true then this is technically the most important thing to say about this property? Hmm.

confused-Techie replied:
It could be? I've already pushed a change to ignore warnings and other notices from the top of documents. So this will no longer show. But we could easily include and even parse warnings more appropriately if we wanted

confused-Techie commented:
Alright, just pushed some changes, now I'll have to correct myself, when I said "nearly all of them" turns out that was a lie lol.
Got the majority in a consecutive block, but truth be told there are still 225 properties without any descriptions.

Doing a spot check on those remaining I am able to find them listed in their respective official spec although not on MDN.

Really all this whole ordeal is teaching me is maybe I should start contributing to MDN for some of these undocumented elements.

If you'd also like to do a spot check on any remaining values the best thing to do is:

1) Find the respective property within the [source](https://github.com/w3c/webref/tree/main/ed/css) file of all properties.
2) Use the `spec.url` at the top of the file that contains your mystery property
3) Browse the specification there to find the relevant properties.

DeeDeeG commented:
Some of the `-webkit` prefixed ones are weirdly forever-prefixed per a "compat" specification: https://compat.spec.whatwg.org/#propdef--webkit-align-items

I am looking through the `update.js` code and the data structures from W3C / MDN stuff and seeing if I can't figure out how to at least grab which properly standard (non-prefixed) property name they are an alias of.

Mostly it's the prefixed name minus the `-webkit-` part, but a few like `-webkit-mask-box-image-outset` are "find `-webkit-mask-box-` replace with `mask-border`".

**Maybe we should programatically generate description strings that just say "alias of: [non-prefixed property name here]"?**

Except for `-webkit-background-size` which is technically slightly different from `background-size`, according to: https://github.com/whatwg/compat/issues/28

confused-Techie commented:
> Some of the `-webkit` prefixed ones are weirdly forever-prefixed per a "compat" specification: https://compat.spec.whatwg.org/#css-simple-aliases
> 
> I am looking through the `update.js` code and the data structures from W3C / MDN stuff and seeing if I can't figure out how to at least grab which properly standard (non-prefixed) property name they are an alias of.
> 
> Mostly it's the prefixed name minus the `-webkit-` part, but a few like `-webkit-mask-box-image-outset` are "find `-webkit-mask-box` replace with `mask-border`".
> 
> **Maybe we should programatically generate description strings that just say "alias of: [non-prefixed property name here]"?**
> 
> Except for `-webkit-background-size` which is technically slightly different from `background-size`, according to: [whatwg/compat#28](https://github.com/whatwg/compat/issues/28)

Yeah I've been looking around for a solid solution here.

The only thing I can see to find is either we can build a manually maintained list based off the resource you provided, or we accept all of those attributes having no description.

Since there are many vendor prefixed items that do have proper descriptions, but many that don't.

At this point it really does make it tempting to try and build a much more intelligent parser for this sort of thing, as it's hard to believe no single resource has the information we need in a machine parsable way.

But which do you think is better then? Leaving as is or a manually updated list, since we could do either. And if a manual list can always default to assuming it's its own valid vendor prefixed item

DeeDeeG commented:
I got this list of remaining properties with empty description `""`:

<details><summary>Properties with empty <code>"description"</code> (click to expand):</summary>

```
% git branch --show-current
update-autocomplete-css
% git rev-parse HEAD
732a353f45f9216133ddc5695361defc25e46e2e
> for (const prop in completions.properties) { if (completions.properties[prop].description === "") { console.log(prop) };  };
```
-webkit-align-items
-webkit-align-content
-webkit-align-self
-webkit-animation-name
-webkit-animation-duration
-webkit-animation-timing-function
-webkit-animation-iteration-count
-webkit-animation-direction
-webkit-animation-play-state
-webkit-animation-delay
-webkit-animation-fill-mode
-webkit-animation
-webkit-backface-visibility
-webkit-background-clip
-webkit-background-origin
-webkit-background-size
-webkit-border-bottom-left-radius
-webkit-border-bottom-right-radius
-webkit-border-top-left-radius
-webkit-border-top-right-radius
-webkit-border-radius
-webkit-box-shadow
-webkit-box-sizing
-webkit-flex
-webkit-flex-basis
-webkit-flex-direction
-webkit-flex-flow
-webkit-flex-grow
-webkit-flex-shrink
-webkit-flex-wrap
-webkit-filter
-webkit-justify-content
-webkit-mask
-webkit-mask-box-image-outset
-webkit-mask-box-image-repeat
-webkit-mask-box-image-slice
-webkit-mask-box-image-source
-webkit-mask-box-image-width
-webkit-mask-clip
-webkit-mask-image
-webkit-mask-origin
-webkit-mask-position
-webkit-mask-repeat
-webkit-mask-size
-webkit-order
-webkit-perspective
-webkit-perspective-origin
-webkit-transform-origin
-webkit-transform-style
-webkit-transform
-webkit-transition-delay
-webkit-transition-duration
-webkit-transition-property
-webkit-transition-timing-function
-webkit-transition
-webkit-text-size-adjust
-webkit-box-align
-webkit-box-flex
-webkit-box-ordinal-group
-webkit-box-orient
-webkit-box-pack
grid-row-gap
grid-column-gap
grid-gap
anchor-scroll
anchor-name
anchor-default
position-fallback
background-position-inline
background-position-block
corner-shape
corners
border-limit
border-clip
border-clip-top
border-clip-right
border-clip-bottom
border-clip-left
margin-break
color-adjust
string-set
bookmark-level
bookmark-label
bookmark-state
reading-order
layout-order
wrap-flow
wrap-through
font-synthesis-weight
font-synthesis-style
font-synthesis-small-caps
copy-into
footnote-display
footnote-policy
object-view-box
baseline-source
text-edge
leading-trim
inline-sizing
initial-letter-wrap
line-grid
line-snap
box-snap
link-parameters
marker-side
spatial-navigation-contain
spatial-navigation-action
spatial-navigation-function
overflow-clip-margin-top
overflow-clip-margin-right
overflow-clip-margin-bottom
overflow-clip-margin-left
overflow-clip-margin-block-start
overflow-clip-margin-inline-start
overflow-clip-margin-block-end
overflow-clip-margin-inline-end
overflow-clip-margin-inline
overflow-clip-margin-block
block-ellipsis
line-clamp
continue
max-lines
float-reference
float-defer
float-offset
page
flow-into
flow-from
region-fragment
block-step-size
block-step-insert
block-step-align
block-step-round
block-step
border-boundary
ruby-merge
ruby-overhang
scroll-start
scroll-start-target
scroll-start-x
scroll-start-y
scroll-start-inline
scroll-start-block
shape-inside
shape-padding
min-intrinsic-sizing
voice-volume
voice-balance
speak
speak-as
pause-before
pause-after
pause
rest-before
rest-after
rest
cue-before
cue-after
cue
voice-family
voice-rate
voice-pitch
voice-range
voice-stress
voice-duration
word-boundary-detection
word-boundary-expansion
text-space-collapse
text-space-trim
hyphenate-limit-zone
hyphenate-limit-lines
hyphenate-limit-last
word-wrap
text-wrap
wrap-before
wrap-after
wrap-inside
text-align-all
text-group-align
line-padding
text-spacing
text-decoration-trim
text-decoration-skip-self
text-decoration-skip-box
text-decoration-skip-spaces
text-emphasis-skip
caret-shape
caret
nav-up
nav-right
nav-down
nav-left
input-security
-webkit-user-select
-webkit-appearance
view-transition-name
fill-break
fill-color
fill-image
fill-origin
fill-position
fill-size
fill-repeat
stroke-align
stroke-break
stroke-dash-corner
stroke-dash-justify
stroke-color
stroke-image
stroke-origin
stroke-position
stroke-size
stroke-repeat
view-timeline-name
view-timeline-axis
view-timeline-inset
view-timeline
animation-range
animation-range-start
animation-range-end
stroke-alignment
stroke-dashcorner
stroke-dashadjust
shape-subtract
marker

---

</details>

The majority are not prefixed.

---

I think there's been a lot of progress made with this PR, though:

```console
% wc -l before_completions.txt after_completions.txt empty_description.txt 
     237 before_completions.txt
     673 after_completions.txt
     225 empty_description.txt
```

With this PR, we have more new properties with descriptions (673 - 225 = **448**) than there used to be completions, period. `autocomplete-css` specs are passing in CI again, now that `d` property has a non-empty `description`.

I'm thinking this is good enough, and if we want to get descriptions for the remaining properties that can be follow-up work? But I will say, with you having taken the initiative to do this, I would want to defer to your opinion about how hard to try and get descriptions for all (or at least most) of the remaining properties. It seems kinda tricky to me, TBH, but I think we could figure out the rest of them _eventually_ at the current pace (especially since a few are from the same spec). It'd just take some time. Maybe a fair amount of time.

DeeDeeG commented:
> At this point it really does make it tempting to try and build a much more intelligent parser for this sort of thing, as it's hard to believe no single resource has the information we need in a machine parsable way.

I dunno, there's this I guess? https://www.w3.org/TR/CSS/#properties

Not super machine-readable, but maybe scrape-able and we can go from there? Maybe the source of this page is checked in to a repo somewhere? Hmm. I see it's checked in here: https://github.com/w3c/csswg-drafts/blob/main/css-2023/Overview.bs#L1045

But there is some magic to it, because it doesn't have anywhere near all the content in that file. I assume a lot is added during the build. I think their CI uses some custom W3C tool called `bikeshed`. https://github.com/w3c/csswg-drafts/blob/main/.github/workflows/build-specs.yml#L54 And they publish the generated artifact(s) of that straight to GitHub Pages' servers is seems like, not saved to a branch, so the final page and source I linked to seem to be it for the formats this resource is published as.

confused-Techie commented:
@DeeDeeG yeah so looking through the properties left I was able to automatically collect one more. But beyond that it seems none of these items exist on MDN to the best of my ability, so limit our ability to collect them automatically. Despite my best efforts in [doing so](https://github.com/mdn/content/pull/25077).

So for the time being for some of these last elements I may just go ahead and build a small manual list of descriptions to include.

Like we have determined having empty descriptions is not the end of the world, so while in the future this list will not be all inclusive, we can still automatically update these without concern. Adding in the newly empty descriptions is just a cherry on top that wouldn't at all be required. So I'll try to submit a PR with those changes soon, and add what I can. But after all of this I may decide to take the time to try and find a way to get specs like this automatically updated elsewhere so that we here can rely on it, and hopefully anywhere else as well. Since it does seem crazy to me that this isn't easily possible as of right now.

icecream17 commented on packages/autocomplete-css/manual-property-desc.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "shape-subtract": {
+    "desc": "The shape-subtract property allows one to exclude part of the content area form the wrapping area. The excluded area is the addition of all the areas defined in a list of CSS basic shapes and/or SVG shapes.",
```
from?

confused-Techie replied:
This is from the spec listed directly below it: `"spec": "https://svgwg.org/svg2-draft/text.html#TextShapeSubtract"`

confused-Techie replied:
The idea is, for this small list of manual descriptions we can list the specification where we got the description directly below it, to keep them updated, as needed.

icecream17 replied:
idk what happened bc I see from in the link...
![image](https://user-images.githubusercontent.com/58114641/223312394-9c8cf8ba-bfca-4f95-8686-ee90b883ae11.png)

confused-Techie replied:
@icecream17 I'm not sure what you mean, that's what's listed above nearly exactly?

If you mean why it's not exact, this description doesn't exist anywhere we can easily machine parse. Meaning I've typed it in manually or copy/pasted some elements. 

That's why it's in the file of `manual-property-desc.json`. It's a collection of descriptions that we can't get automatically and fill in manually if we would like. This is a result of the above conversation asking what we can do about all the properties with empty descriptions. And the only answer, apart from building a complex parsing machanism that can parse the dozens of spec webpages and websites, is to make a part of our list manually like I'm doing now

confused-Techie replied:
Oh. You're saying that `from` is mispelled. I've fixed it.

icecream17 replied:
oh lmao sry for being very unclear

confused-Techie commented:
@DeeDeeG alright, so I've tried my best to curb the issue of many empty property descriptions.

Like mentioned above the best way I could see us doing this, is unfortunately a manual list. So I've gone ahead and implemented this behavior, where as a last resort our manual list is checked for a property description. 

Once the `update.js` runs it will actually warn you how many empty descriptions you have. With an optional flag of `--show-empty` being available to print every single property with a missing description.

---

I've now added many manual descriptions either copy/pasted from the source spec or manually written using the source spec as a guide to do so. This has now gotten our empty completion descriptions down from `225` to `90`. So big wins there, but I think I've read through as many CSS specs as I can for the day. So either, I'd be more than happy to go ahead and merge this as is, and at a later date anyone can pick this up to finish the descriptions that need to be made manually, or if we really do want them included I can try to get to this again soon, or really anyone else can as well, comitting directly to my branch. But let me know what you think.

<details><summary>The remaining empty description completions</summary>

margin-break
color-adjust
string-set
bookmark-level
bookmark-label
bookmark-state
reading-order
layout-order
wrap-flow
wrap-through
font-synthesis-weight
font-synthesis-style
font-synthesis-small-caps
copy-into
footnote-display
footnote-policy
object-view-box
baseline-source
text-edge
leading-trim
inline-sizing
initial-letter-wrap
line-grid
line-snap
box-snap
link-parameters
marker-side
spatial-navigation-contain
spatial-navigation-action
spatial-navigation-function
overflow-clip-margin-top
overflow-clip-margin-right
overflow-clip-margin-bottom
overflow-clip-margin-left
overflow-clip-margin-block-start
overflow-clip-margin-inline-start
overflow-clip-margin-block-end
overflow-clip-margin-inline-end
overflow-clip-margin-inline
overflow-clip-margin-block
block-ellipsis
line-clamp
continue
max-lines
float-reference
float-defer
float-offset
page
flow-into
flow-from
region-fragment
block-step-size
block-step-insert
block-step-align
block-step-round
block-step
border-boundary
ruby-merge
ruby-overhang
scroll-start
scroll-start-target
scroll-start-x
scroll-start-y
scroll-start-inline
scroll-start-block
shape-inside
shape-padding
min-intrinsic-sizing
text-space-collapse
text-decoration-trim
text-decoration-skip-self
text-decoration-skip-box
text-decoration-skip-spaces
text-emphasis-skip
caret-shape
caret
nav-up
nav-right
nav-down
nav-left
input-security
view-transition-name
fill-color
view-timeline-name
view-timeline-axis
view-timeline-inset
view-timeline
animation-range
animation-range-start
animation-range-end
</details>

confused-Techie commented:
@DeeDeeG just an fyi, I've now updated `manual-property-desc.json` to include every single missing description.

So there should be zero empty descriptions now.

DeeDeeG commented on packages/autocomplete-css/update.js:
```diff
@@ -0,0 +1,366 @@
+    }
+  }
+
+  console.log("Updated all `autocomplete-css` completions.");
+  console.log(`Total Completion Properties without a description: ${count}!`);
+  console.log("It is not required to fix the above empty completions issue.");
+  console.log("Use `node update.js --show-empty` to show the empty property names.");
```
Some of these console.log strings are not super relevant, now that there are zero properties with empty descriptions.

Especially the line "It is not required to fix the above empty completions issue." -- now there is no such issue!

This suggestion is to conditionally not print some of the info if the missing descriptions count is zero.

```suggestion
  console.log("Updated all `autocomplete-css` completions.");
  console.log(`Total Completion Properties without a description: ${count}!`);
  if (count !== 0) {
    console.log("It is not required to fix the above empty completions issue.");
    console.log("Use `node update.js --show-empty` to show the empty property names.");
  };
```


Before:

```
$ node update.js
Updated all `autocomplete-css` completions.
Total Completion Properties without a description: 0!
It is not required to fix the above empty completions issue.
Use `node update.js --show-empty` to show the empty property names.
```

After:

```
$ node update.js
Updated all `autocomplete-css` completions.
Total Completion Properties without a description: 0!
```

confused-Techie replied:
Good call on making this last bit optional, way less verbose when it's not needed

DeeDeeG reviewed (approved):
Okay, thank you for this huge effort. Now with all the descriptions filled out, there's nothing left to object to here, IMO.

I tested the latest binary from this PR, and it's working for me still 👍.

Tried doing `node update.js` and got equal lines added/removed, looks like the order just shuffled around (I assume this might be an OS thing? I'm on Linux right now). So, completions appear to be stable (other than some descriptions were slightly updated, maybe from the mdn repo).

I see there are indeed no properties without descriptions. Thank you again for this huge effort.

I have one comment below about a cosmetic/docs type suggestion of which stuff to log when running `node update.js`. I don't care strongly about that, it is just something I noticed, feel free to apply it or ignore it/decline to apply it.

Approved for the above reasons, and as mentioned in earlier approvals in the thread.

confused-Techie commented:
Thanks a ton to everyone for all the help on this one, and to @DeeDeeG for being so persistent with making sure this got reviewed and had an eye kept on it. Appreciate all of you, and with the above approval I'll go ahead and merge!
